File: sdk/typescript/src/resources/bundles.ts
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0

import { HttpClient } from "../internal/http";
import {
  Bundle,
  BundleWithSummary,
  ClickStats,
  CreateBundleBody,
  Link,
  TimelineRange,
  UpdateBundleBody,
} from "../models";

export class BundlesResource {
  constructor(private readonly http: HttpClient) {}

  /** Get a bundle by ID with aggregated click summary. */
  get(id: number, options: { range?: TimelineRange } = {}): Promise<BundleWithSummary> {
    return this.http.request("GET", `/api/bundles/${id}`, {
      query: { range: options.range },
    });
  }

  /** List bundles. Filter by archived status and click-count range. */
  list(
    options: { archived?: "1" | "true" | "only" | "all"; range?: TimelineRange } = {},
  ): Promise<BundleWithSummary[]> {
    return this.http.request("GET", "/api/bundles", {
      query: { archived: options.archived, range: options.range },
    });
  }

  /** Create a new bundle. */
  create(body: CreateBundleBody): Promise<Bundle> {
    return this.http.request("POST", "/api/bundles", { body });
  }

  /** Update a bundle's name, description, icon, or accent. */
  update(id: number, body: UpdateBundleBody): Promise<Bundle> {
    return this.http.request("PUT", `/api/bundles/${id}`, { body });
  }

  /** Permanently delete a bundle. */
  delete(id: number): Promise<{ deleted: boolean }> {
    return this.http.request("DELETE", `/api/bundles/${id}`);
  }

  /** Archive a bundle. */
  archive(id: number): Promise<Bundle> {
    return this.http.request("POST", `/api/bundles/${id}/archive`);
  }

  /** Unarchive a bundle. */
  unarchive(id: number): Promise<Bundle> {
    return this.http.request("POST", `/api/bundles/${id}/unarchive`);
  }

  /** Get click analytics for a bundle. */
  analytics(id: number, options: { range?: TimelineRange } = {}): Promise<ClickStats> {
    return this.http.request("GET", `/api/bundles/${id}/analytics`, {
      query: { range: options.range },
    });
  }

  /** List links in a bundle. */
  links(id: number): Promise<Link[]> {
    return this.http.request("GET", `/api/bundles/${id}/links`);
  }

  /** Add a link to a bundle. */
  addLink(id: number, linkId: number): Promise<{ added: boolean }> {
    return this.http.request("POST", `/api/bundles/${id}/links`, { body: { linkId } });
  }

  /** Remove a link from a bundle. */
  removeLink(id: number, linkId: number): Promise<{ removed: boolean }> {
    return this.http.request("DELETE", `/api/bundles/${id}/links/${linkId}`);
  }
}

File: sdk/dart/lib/src/resources/bundles.dart
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0

import '../base_client.dart';
import '../models.dart';

/// Methods for the `/api/bundles` and related endpoints.
class BundlesResource {
  BundlesResource(this._http);

  final ShrtnrBaseClient _http;

  /// Get a bundle by ID with aggregated click summary.
  /// Optional [range] scopes the click window.
  Future<BundleWithSummary> get(int id, {String? range}) async {
    final json = await _http.requestJson(
      'GET',
      '/_/api/bundles/$id',
      query: {'range': range},
    );
    return BundleWithSummary.fromJson(json! as Map<String, dynamic>);
  }

  /// List bundles. Use [archived] to filter archived status
  /// (`'1'`, `'true'`, `'only'`, `'all'`). Optional [range] scopes click counts.
  Future<List<BundleWithSummary>> list({String? archived, String? range}) async {
    final json = await _http.requestJson(
      'GET',
      '/_/api/bundles',
      query: {'archived': archived, 'range': range},
    );
    return (json! as List<dynamic>)
        .map((dynamic e) =>
            BundleWithSummary.fromJson(e as Map<String, dynamic>))
        .toList(growable: false);
  }

  /// Create a new bundle.
  Future<Bundle> create({
    required String name,
    String? description,
    String? icon,
    String? accent,
  }) async {
    final body = <String, dynamic>{'name': name};
    if (description != null) body['description'] = description;
    if (icon != null) body['icon'] = icon;
    if (accent != null) body['accent'] = accent;
    final json = await _http.requestJson('POST', '/_/api/bundles', body: body);
    return Bundle.fromJson(json! as Map<String, dynamic>);
  }

  /// Update a bundle's name, description, icon, or accent.
  Future<Bundle> update(
    int id, {
    String? name,
    String? description,
    String? icon,
    String? accent,
  }) async {
    final body = <String, dynamic>{};
    if (name != null) body['name'] = name;
    if (description != null) body['description'] = description;
    if (icon != null) body['icon'] = icon;
    if (accent != null) body['accent'] = accent;
    final json =
        await _http.requestJson('PUT', '/_/api/bundles/$id', body: body);
    return Bundle.fromJson(json! as Map<String, dynamic>);
  }

  /// Permanently delete a bundle. Member links are preserved.
  Future<DeletedResult> delete(int id) async {
    final json = await _http.requestJson('DELETE', '/_/api/bundles/$id');
    return DeletedResult.fromJson(json! as Map<String, dynamic>);
  }

  /// Archive a bundle. It stays in the database but is hidden from the
  /// default listing.
  Future<Bundle> archive(int id) async {
    final json = await _http.requestJson('POST', '/_/api/bundles/$id/archive');
    return Bundle.fromJson(json! as Map<String, dynamic>);
  }

  /// Restore a previously archived bundle.
  Future<Bundle> unarchive(int id) async {
    final json =
        await _http.requestJson('POST', '/_/api/bundles/$id/unarchive');
    return Bundle.fromJson(json! as Map<String, dynamic>);
  }

  /// Get combined click analytics for a bundle. Optional [range] scopes the window.
  Future<ClickStats> analytics(int id, {String? range}) async {
    final json = await _http.requestJson(
      'GET',
      '/_/api/bundles/$id/analytics',
      query: {'range': range},
    );
    return ClickStats.fromJson(json! as Map<String, dynamic>);
  }

  /// List links in a bundle.
  Future<List<Link>> links(int id) async {
    final json = await _http.requestJson('GET', '/_/api/bundles/$id/links');
    return (json! as List<dynamic>)
        .map((dynamic e) => Link.fromJson(e as Map<String, dynamic>))
        .toList(growable: false);
  }

  /// Add a link to a bundle. Idempotent.
  Future<AddedResult> addLink(int id, int linkId) async {
    final json = await _http.requestJson(
      'POST',
      '/_/api/bundles/$id/links',
      body: <String, dynamic>{'link_id': linkId},
    );
    return AddedResult.fromJson(json! as Map<String, dynamic>);
  }

  /// Remove a link from a bundle. The link itself is not deleted.
  Future<RemovedResult> removeLink(int id, int linkId) async {
    final json =
        await _http.requestJson('DELETE', '/_/api/bundles/$id/links/$linkId');
    return RemovedResult.fromJson(json! as Map<String, dynamic>);
  }
}
